shader_recompiler/LOP3: Use brute force python results within switch/case.

degasus opened the pull request:
Thanks to @asLody for optimizing this function within #7985. This raised the focus that this function should be optimized more.

The current table assumes that the host GPU is able to invert for free, so only AND,OR,XOR are accumulated in the performance metrik.

Performance results:

Instructions
0: 8
1: 30
2: 114
3: 80
4: 24

Latency
0: 8
1: 30
2: 194
3: 24

lioncash commented on src/shader_recompiler/frontend/maxwell/translate/impl/logic_operation_three_input.cpp:
```diff
@@ -14,56 +14,794 @@ namespace {
+    switch (ttbl) {
+    // automatically generated code, do not edit manually
+    case 0:
+        r = ir.Imm32(0);
```
Given how nothing is done after the switch, you can make this slightly more compact by returning in the case instead of assigning then breaking out.

degasus replied:
done

Morph1984 commented:
Supersedes #7985

v1993 commented on src/shader_recompiler/frontend/maxwell/translate/impl/logic_operation_three_input_lut3.py:
```diff
@@ -0,0 +1,79 @@
+s = """    case {imm}:
+        return {op};"""
+for imm in range(256):
+    print(s.format(imm=imm, op=optimized_calls[imm][0]))
```
Missing newline at EOF

merryhime commented:
I note a lot of the cases could be improved by assigning a cost to `not`, e.g.: For 0x08 I would probably prefer `ir.BitwiseAnd(ir.BitwiseNot(a), ir.BitwiseAnd(b, c)));`, but *shrug*.

degasus commented:
@merryhime I've added support for variable amount of arguments to the instructions and added cost(NEG) = 0.1. So it is now trying to minimize the latency and the amount if NEG calls as long as it doesn't increase the cost of AND,OR,XOR.
In case there are still two with equal cost, it prefers the one with the shorter code string.

merryhime commented:
@degasus NPN classification is the idea that given three transformations (**n**egating the output / **p**ermuting the inputs / **n**egating the inputs), you can classify luts into families of related circuits; it's a well known result that there are 14 such families for 3-luts.
